tutorials/week4-react.md

```diff
@@ -0,0 +1,689 @@
+        counter.count += counter.increment;
+        setCounter(counter); // will not cause component to rerender
+    }
+
+  // Will not get called when incrementCount is called
+    useEffect(() => {
+        console.log(`The current count is ${counter.count}`);
+    }, [counter]);
+
+    return (
+        <div>
+            <h1>Count: {count}</h1>
+            <button onClick={incrementCount}>Click me!</button>
+        </div>
+    );
+}
+
+export default Counter;
+```
+In the example above, we've modified the counter to be an object containing both the current count and the amount to increment the counter by when the button is clicked.
+
+Similar to the list case for useState, updating an attribute of an object and setting the state variable to that same object reference will not cause a re-render, and useEffect will not detect a change in its dependencies, even though we have changed one of the attributes. To properly invoke a useEffect call, we must change the object reference itself:
+
+```tsx
+function Counter() {
+    const [counter, setCounter] = useState({ count: 0, increment: 1 });
+
+    function incrementCount() {
+      // we use the spread operator (...) to make a copy of the object
+        setCounter({...counter, count: counter.count + counter.increment}); 
+    }
+
+    useEffect(() => {
+        console.log(`The current count is ${counter.count}`);
+    }, [counter]);
+
+    return (
+        <div>
+            <h1>Count: {count}</h1>
+            <button onClick={incrementCount}>Click me!</button>
+        </div>
+    );
+}
+
+export default Counter;
+```
+
+The useEffect hook will now be invoked with each button click. However, our useEffect is dependent on the entire `counter` object, meaning that a change to any of its attributes will cause the useEffect hook to be called, which is not always a desired behavior. We can get around this by being more specific in our dependency list: 
+
+```tsx
+function Counter() {
+    const [counter, setCounter] = useState({ count: 0, increment: 1 });
+
+    function incrementCount() {
+      // we use the spread operator (...) to make a copy of the object
+        setCounter({...counter, count: counter.count + counter.increment}); 
+    }
+
+    function incrementIncrement() {
+      // we use the spread operator (...) to make a copy of the object
+        setCounter({...counter, increment: counter.increment + 1}); 
+    }
+
+    useEffect(() => {
+        console.log(`The current count is ${counter.count}`);
+    }, [counter.count]); // Now that we only depend on count, this function will not be called if the incrementIncrement function is called
+
+    return (
+        <div>
+            <h1>Count: {count}</h1>
+            <button onClick={incrementCount}>Click me to increment the count!</button>
+            <button onClick={incrementIncrement}>Click me to increase the increment amount!</button>
+        </div>
+    );
+}
+
+export default Counter;
+```
```

Comment:
Maybe we can add this point at the end to encourage a read on other react hooks as well.

Feel free to explore additional React hooks like `useContext`, `useCallback`, `useMemo`, and `useReducer`, as they are essential tools for efficient React development.